### PR TITLE
Issue #1940 Add unit tests for addon apply/remove actions

### LINKED ISSUE #1940: Add unit tests for Apply and Remove addon functions
Add unit tests for `AddonApply` and `AddonRemove` functions. 
https://github.com/minishift/minishift/blob/master/pkg/minishift/addon/manager/addon_manager.go#L195-L261

### PR DESCRIPTION
**Pull request by budhrg**
Fix #1940

### CODE REVIEW

File: test/testdata/testaddons/invalidaddon/invalidaddon.addon [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Name: invalidaddon
+# Description: Test addon description
+# URL: http://testaddon.url
+# Required-Vars: TEST
+
+echo This testaddon is having variable TEST with #{TEST} value
```
Comment (budhrg) [inline test/testdata/testaddons/invalidaddon/invalidaddon.addon]: This is intended here. We don't require new line here.

**Comment by anjannath** [comment]
Upon running the test i get an output like this:
```
--- PASS: TestApplyAddon (0.00s)
=== RUN   TestRemoveAddon
-- Removing addon 'testaddon':
Removing testaddon with variable TEST of foo value

--- PASS: TestRemoveAddon (0.00s)
=== RUN   TestApplyInvalidAddon
-- Applying addon 'invalidaddon':--- PASS: TestApplyInvalidAddon (0.00s)
=== RUN   TestRemoveInvalidAddon
-- Removing addon 'invalidaddon':--- PASS: TestRemoveInvalidAddon (0.00s)
PASS
ok  	github.com/minishift/minishift/pkg/minishift/addon/manager	0.026s
```
_TestApplyInvalidAddon_ and _TestRemoveInvalidAddon_ has weired output, there's no new line between the name of the unit test and test result. This happens because we error out early as the addon is an invalid one, and don't even reach this [line](https://github.com/minishift/minishift/blob/master/pkg/minishift/addon/manager/addon_manager.go#L225). Should probably add a new line after this [line](https://github.com/minishift/minishift/blob/master/pkg/minishift/addon/manager/addon_manager.go#L196) in _addon_manager.go_

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 2]
```diff
@@ -182,6 +185,109 @@ func TestVerifyMissingRequiredVariablesInContext(t *testing.T) {
 	}
 }
 
+type FakeSSHDockerCommander struct {
+	docker.DockerCommander
+	provision.SSHCommander
```
Comment (LalatenduMohanty) [inline pkg/minishift/addon/manager/addon_manager_test.go]: @budhrg Doesn't need something like this?
```

	ocRunner             *oc.OcRunner
	dockerCommander      docker.DockerCommander
	sshCommander         provision.SSHCommander
	interpolationContext InterpolationContext
```

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 3]
```diff
@@ -182,6 +185,109 @@ func TestVerifyMissingRequiredVariablesInContext(t *testing.T) {
...
+}
+
+func (f *FakeSSHDockerCommander) SSHCommand(args string) (string, error) {
+	return "openshift v3.6.1+008f2d5\nkubernetes v1.6.1+5115d708d7\netcd 3.2.1", nil
+}
```
[the remark this answers is not in the document]
Comment (budhrg) [reply pkg/minishift/addon/manager/addon_manager_test.go]: Both AddonApply/Remove functions are using `verifyRequiredOpenshiftVersion` which has ssh command execution to get OpenShift version within the VM.

See https://github.com/minishift/minishift/blob/b4187fc1430b2347dfbd406d64875b7c4f72638d/pkg/minishift/addon/manager/addon_manager.go#L301

And, in the https://github.com/minishift/minishift/blob/b4187fc1430b2347dfbd406d64875b7c4f72638d/pkg/minishift/docker/docker.go#L149-L153 
implementation, it is the `SSHCommand` being invoked hence, only mocking this function will make sure that we don't require info from VM.

File: pkg/minishift/addon/manager/addon_manager_test.go (hunk 2 above)
Comment (budhrg) [reply pkg/minishift/addon/manager/addon_manager_test.go]: See https://github.com/minishift/minishift/pull/1941#discussion_r164434817

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 4]
```diff
@@ -182,6 +185,109 @@ func TestVerifyMissingRequiredVariablesInContext(t *testing.T) {
...
+
+var expectedApplyAddonOutput = `-- Applying addon 'testaddon':
+This testaddon is having variable TEST with foo value
+
+`
```
[the remark this answers is not in the document]
Comment (budhrg) [reply pkg/minishift/addon/manager/addon_manager_test.go]: Fixed

**Comment by budhrg** [comment]
retest this please

**Comment by budhrg** [comment]
Rebased with upstream master after https://github.com/minishift/minishift/pull/1755 got merged.

File: pkg/minishift/addon/manager/addon_manager_test.go [hunk 5]
```diff
@@ -146,6 +150,81 @@ func TestVerifyMissingRequiredVariablesInContext(t *testing.T) {
...
+	path := filepath.Join(basepath, "..", "..", "..", "..", "test", "testdata", "testaddons")
+	manager, err := NewAddOnManager(path, make(map[string]*addon.AddOnConfig))
+	assert.NoError(t, err, "Unexpected error creating manager in directory '%s'", path)
+
+	testaddon := manager.Get("invalidaddon")
+	context, err := command.NewExecutionContext(nil, &FakeSSHDockerCommander{})
+	assert.NoError(t, err, "Unexpected error creating new execution context")
+
+	err = manager.ApplyAddOn(testaddon, context)
+	assert.EqualError(t, expectedInvalidAddonOperationError, err.Error())
```
Comment (budhrg) [inline pkg/minishift/addon/manager/addon_manager_test.go]: Looking nice here :smile:

Comment (anjannath) [reply pkg/minishift/addon/manager/addon_manager_test.go]: Indeed! 😄

**Pull request merged**